1-Draft/RFCNNNN-Download-Cmdlet.md

```diff
@@ -0,0 +1,294 @@
+    Uri GetProxy();
+    void SetProxyCredential(PSCredential Credential);
+    PSCredential GetProxyCredential();
+}
+
+internal interface IInvokeDownloadRedirectable
+{
+    IList<Uri> GetRedirectionChain(PSCmdlet Cmdlet, Uri RequestUri);
+}
+```
+
+Class implementing these interfaces will be registered to a static internal dictionary
+
+```csharp
+public class InvokeDownloadCmdlet : PSCmdlet
+{
+    internal static Dictionary<String, IInvokeDownloadProtocol> ProtocolHandlers;
+}
+```
+
+These APIs will begin as internal for several versions.
+Once they are mature, they will be made public.
+This will allow users to extend and replace protocols with their own handlers.
+A set of default handlers will be initialize.
+The initial default protocols will be `HTTP`, `HTTPS`, `FTP`, and `FILE`.
+
+The protocol handler chosen will be based on the `Uri.Scheme`.
+When a URI does not include a scheme, `HTTPS` will be assumed.
+In the event that the supplied scheme has not been registered, an error will be thrown.
+
+The protocol handlers are primarily responsible for returning a `Stream`.
+The `Stream` will be used by the cmdlet.
+The provided `Stream` will be read from and the results will be written to the file.
+
+`IInvokeDownloadRedirectable` handlers indicate that the protocol supports redirection.
+They are responsible for following the redirection chain and returning the URIs in that chain.
+The cmdlet will use the final URI to perform the download and resolve the file name if required.
+
+### Resume Feature
+
+For protocols that support resuming failed or interrupted file transfer,
+the protocol handler can implement `IInvokeDownloadResumable`.
+
+The protocol handler will be responsible for its own resume implementation.
+The protocol handler will indicate if resume was successful.
+The protocol handler will indicate if the resume is already complete,
+i.e. the file is already completely downloaded and there is nothing to resume.
+
+If the resume was successful, the cmdlet will append to existing file.
+If resume was not successful, the cmdlet will attempt a full download load of the same file.
+If the resume is already complete, no action will be taken.
+
+## Alternate Proposals and Considerations
+
+Instead of being extensible, the cmdlet could be a closed system.
+There have been many lessons learned from the Web Cmdlets with regards to extensibility.
+Adding features is made difficult when all the code is too tightly coupled.
+
+Additional Features currently present in the Web Cmdlet might be considered.
+`-Headers` for example may be useful.
+However, the goal should be to keep the cmdlet focused on simple downloads.
+Complex download scenarios can still be handled by the Web Cmdlets.
+Also, the cmdlet should be considered as protocol agnostic as possible.
+`-Headers` may make sense for `HTTP` and `HTTPS`,
+but they will make no sense for `SSH`.
+
+Alternate automated file name resolution strategies could be considered.
```

Comment:
Generally, PowerShell's cmdlets simply _fail_ if the target file already exists and require `-Force` to force replacement; examples: `Move-Item`, `New-Item`.

Reply:
I think in this one instance it is better to have parity with peer technologies than with PowerShell. There are decades worth of expected behavior built into this kind of functionality, seeking functional parity with PowerShell in this case may not be the best approach.

Reply:
Should still have a `-Force` to force replacement, right?

Reply:
And just to clarify, if the `-Path` is specified and the file already exists, then we would get the error and have to use a `-Force` to save it. The automated name only applied when the `-Path` is not specified.

Reply:
another discussion point on Path which would break what you propose here is whether a folder can be specified and allowing automatic name resolution create the file in the designated folder.

```powershell
Invoke-Download 'http://contoso.com/file.exe' -Path `c:\downloads\`
```

Reply:
On `-Force`, I cringe at it, not the idea of allow overwriting the file, just the vague parameter name. it works well when the verb and noun make it clear what `-Force` would do. But the `get-*` commands that have `-Force` are mostly about showing hidden items, and there are no `Invoke-*`  commands that have `-Force`.

I guess one assumption would be that if you are using `-Force` you would already know the file exists? Otherwise, would you really care about the autogen file name enough to overwrite if you didn't? And if you did know it exists, you could supply a path.

Again, not opposed to the idea. Just need to feel out all the possible issues with it.

Reply:
My expectation when calling with a `-Path` to a file is that I would be able to save the file to that path. 

    Invoke-Download 'http://contoso.com/layout.csv' -Path `c:\website\layout.csv`

If this will not overwrite the target file, then I would expect to be able to call `-Force`.

    Invoke-Download 'http://contoso.com/layout.csv' -Path `c:\website\layout.csv` -Force

I would use `-Force` to not care if the file already exists.

Reply:
That's a good point.. I'm too used to the Web Cmdlets which implemented `-OutFile` instead of `-Path`/`-LiteralPath` and it overwrites if it exists. I'm not sure which is better.. making it work like the Web Cmdlets, more like peer technology, or more like other PowerShell commands... it seems like no matter what behavior is chosen it will be surprising depending on what you are coming from.

Reply:
Prefer `-AllowClobber` to the generic `-Force` which is way overloaded